### PR TITLE
Proof of payment by only revealing key derivation, not the actual tx secret key

### PR DESCRIPTION
**Pull request by stoffu**
This is a refined version of PR #1943. Its follow-up PR #1945 was taken down after a private discussion with @luigi1111.

```
[wallet 42vPNJ]: get_tx_proof 23d4a5075d7272f305036103da70526d224c083ac73c2756e80049e6aad3214f 44AFFq5kSiGBoZ4NMDwYtN18obc8AemS33DBLWs3H7otXft3XjrpDtQGv7SqSsaBYBb98uNbr2VBBEt7f2wfn3RVGQBEP3A
Signature: ProofV1ihLMTQQFuaoH6GiBnu5PzZ2hhocmMZDMpHQWEXLB7HqBKNYXFbUASND7k3J3hCAgk2NocW2jxQ3Ngd4vWGCez1pFXMUc5bt7snVfiWH2ohzF1H7m1RThkuVZqRvYASMENuni

[wallet 42vPNJ]: check_tx_proof 23d4a5075d7272f305036103da70526d224c083ac73c2756e80049e6aad3214f 44AFFq5kSiGBoZ4NMDwYtN18obc8AemS33DBLWs3H7otXft3XjrpDtQGv7SqSsaBYBb98uNbr2VBBEt7f2wfn3RVGQBEP3A ProofV1ihLMTQQFuaoH6GiBnu5PzZ2hhocmMZDMpHQWEXLB7HqBKNYXFbUASND7k3J3hCAgk2NocW2jxQ3Ngd4vWGCez1pFXMUc5bt7snVfiWH2ohzF1H7m1RThkuVZqRvYASMENuni
Good signature
44AFFq5kSiGBoZ4NMDwYtN18obc8AemS33DBLWs3H7otXft3XjrpDtQGv7SqSsaBYBb98uNbr2VBBEt7f2wfn3RVGQBEP3A received 0.200000000000 in txid <23d4a5075d7272f305036103da70526d224c083ac73c2756e80049e6aad3214f>
This transaction has 45803 confirmations
```

### CODE REVIEW

**Comment by luigi1111** [comment]
After thinking about this (again), I think adding txkey verification is desirable.

Something like:

```
txsec, txpub
B,A (address)
D = txsec*A (derivation)

comm {
msg,
D, //committing to A is unnecessary
L1,
R1
}

GEN
L1 = kG
R1 = kA
c = Hs(comm)
r = k - c*txsec

VER
L1 = rG + cR
R1 = rA + cD
Hs(comm) == c
```

The reason is, "base" A isn't _really_ committed to, and once D is revealed, anyone can forge signatures for other A's for the same B (not pre-chosen A's, however, just "random" ones) without knowledge of txsec.


There was a second issue where anyone (without the proposed change above) or those with knowledge of txsec (with the proposed change) could create random new address which "match" the output(s) in question. However, RingCT make this issue moot, because it actually uses the derivation to decrypt amount data; if an output pubkey matches, but the derivation fails to decode valid RCT amount data, you can deduce that either: 1. the output data is corrupt or 2. the address is a fake.

**Comment by stoffu** [comment]
Thanks @luigi1111 for reviewing! The code is updated accordingly.

```
[wallet 42vPNJ]: get_tx_proof 23d4a5075d7272f305036103da70526d224c083ac73c2756e80049e6aad3214f 44AFFq5kSiGBoZ4NMDwYtN18obc8AemS33DBLWs3H7otXft3XjrpDtQGv7SqSsaBYBb98uNbr2VBBEt7f2wfn3RVGQBEP3A
Wallet password: 
Tx key: bd818e47f72fdd5a16f8c1125b963940c2b933f30cfe8772f794e7d81f63da00
Signature: SigV2ihLMTQQFuaoH6GiBnu5PzZ2hhocmMZDMpHQWEXLB7HqBSLRni49vYFCUvX9qLvVfszZkkCqHNVaf8MhtxxHzvYGsf5fsduf2ChchLXAPKT8dNMNGfNay1h5LQEJ98n3ZrmYb
[wallet 42vPNJ]: check_tx_proof 23d4a5075d7272f305036103da70526d224c083ac73c2756e80049e6aad3214f 44AFFq5kSiGBoZ4NMDwYtN18obc8AemS33DBLWs3H7otXft3XjrpDtQGv7SqSsaBYBb98uNbr2VBBEt7f2wfn3RVGQBEP3A SigV2ihLMTQQFuaoH6GiBnu5PzZ2hhocmMZDMpHQWEXLB7HqBSLRni49vYFCUvX9qLvVfszZkkCqHNVaf8MhtxxHzvYGsf5fsduf2ChchLXAPKT8dNMNGfNay1h5LQEJ98n3ZrmYb
Good signature
44AFFq5kSiGBoZ4NMDwYtN18obc8AemS33DBLWs3H7otXft3XjrpDtQGv7SqSsaBYBb98uNbr2VBBEt7f2wfn3RVGQBEP3A received 0.200000000000 in txid <23d4a5075d7272f305036103da70526d224c083ac73c2756e80049e6aad3214f>
This transaction has 31721 confirmations
```

File: src/simplewallet/simplewallet.cpp [hunk 1]
```diff
@@ -3209,6 +3212,76 @@ bool simple_wallet::get_tx_key(const std::vector<std::string> &args_)
...
+    fail_msg_writer() << tr("failed to parse address");
+    return true;
+  }
+
+  LOCK_IDLE_SCOPE();
+
+  crypto::secret_key tx_key;
+  if (m_wallet->get_tx_key(txid, tx_key))
+  {
+    success_msg_writer() << tr("Tx key: ") << epee::string_tools::pod_to_hex(tx_key);
```
Comment (moneromooo-monero) [inline src/simplewallet/simplewallet.cpp]: Pedantic, but I'd use message_writer for non fatal/success messages.

File: src/simplewallet/simplewallet.cpp [hunk 2]
```diff
@@ -3209,6 +3212,76 @@ bool simple_wallet::get_tx_key(const std::vector<std::string> &args_)
...
+      return true;
+    }
+    cryptonote::blobdata tx_key_data;
+    if(!epee::string_tools::parse_hexstr_to_binbuff(args[2], tx_key_data) || tx_key_data.size() != sizeof(crypto::secret_key))
+    {
+      fail_msg_writer() << tr("failed to parse tx_key");
+      return true;
+    }
+    tx_key = *reinterpret_cast<const crypto::secret_key*>(tx_key_data.data());
+  }
```
Comment (moneromooo-monero) [inline src/simplewallet/simplewallet.cpp]: This seems to accept a tx key as last parameter, but silently ignore it if one is found in the wallet cache

File: src/simplewallet/simplewallet.cpp [hunk 3]
```diff
@@ -3209,6 +3212,76 @@ bool simple_wallet::get_tx_key(const std::vector<std::string> &args_)
...
+  {
+    crypto::generate_tx_proof(txid, R, address.m_view_public_key, rA, tx_key, sig);
+  }
+  catch (const std::runtime_error &e)
+  {
+    fail_msg_writer() << e.what();
+    return true;
+  }
+
+  std::string sig_str = std::string("SigV2") +
```
Comment (moneromooo-monero) [inline src/simplewallet/simplewallet.cpp]: I'd rather you use another prefix, since SigV2 would be used in case we need to change the way message signatures work from SigV1. Maybe ProofV1 or similar ?

File: src/simplewallet/simplewallet.cpp [hunk 4]
```diff
@@ -3381,6 +3451,129 @@ bool simple_wallet::check_tx_key(const std::vector<std::string> &args_)
...
+    success_msg_writer() << tr("Good signature");
+  }
+  else
+  {
+    fail_msg_writer() << tr("Bad signature");
+    return true;
+  }
+
+  // obtain key derivation by multiplying scalar 1 to the pubkey r*A included in the signature
+  crypto::secret_key one = AUTO_VAL_INIT(one);
```
Comment (moneromooo-monero) [inline src/simplewallet/simplewallet.cpp]: Does this set to 0 ? If so, is this a standard thing ? I don't recall any particular ctor. Otherwise, there is null_skey IIRC (or rct::I).

File: src/simplewallet/simplewallet.cpp [hunk 5]
```diff
@@ -3381,6 +3451,129 @@ bool simple_wallet::check_tx_key(const std::vector<std::string> &args_)
...
+  {
+    fail_msg_writer() << tr("failed to validate transaction from daemon");
+    return true;
+  }
+  if (tx_hash != txid)
+  {
+    fail_msg_writer() << tr("failed to get the right transaction from daemon");
+    return true;
+  }
+  crypto::public_key R = get_tx_pub_key_from_extra(tx);
```
Comment (moneromooo-monero) [inline src/simplewallet/simplewallet.cpp]: This can return null_pkey if not found (not sure if the following code would assert on invalid key, or just say no money found - the latter would be ok, though it's clearer to error out here)

File: src/simplewallet/simplewallet.cpp [hunk 6]
```diff
@@ -3381,6 +3451,129 @@ bool simple_wallet::check_tx_key(const std::vector<std::string> &args_)
...
+  if(args.size() != 3) {
+    fail_msg_writer() << tr("usage: check_tx_proof <txid> <address> <signature>");
+    return true;
+  }
+
+  if (!try_connect_to_daemon())
+    return true;
+
+  // parse txid
+  assert(m_wallet);
```
Comment (moneromooo-monero) [inline src/simplewallet/simplewallet.cpp]: Maybe error out, insead of crashing on purpose. Not a great issue here since it's the wallet from user commandthough.

File: src/simplewallet/simplewallet.cpp [hunk 7]
```diff
@@ -3381,6 +3451,129 @@ bool simple_wallet::check_tx_key(const std::vector<std::string> &args_)
...
+  cryptonote::blobdata txid_data;
+  if(!epee::string_tools::parse_hexstr_to_binbuff(args[0], txid_data) || txid_data.size() != sizeof(crypto::hash))
+  {
+    fail_msg_writer() << tr("failed to parse txid");
+    return true;
+  }
+  crypto::hash txid = *reinterpret_cast<const crypto::hash*>(txid_data.data());
+
+  LOCK_IDLE_SCOPE();
```
Comment (moneromooo-monero) [inline src/simplewallet/simplewallet.cpp]: This can probably be moved to below, as this may be a lengthy block till the wallet finished its current 1k block chunk.

File: src/crypto/crypto.cpp [hunk 8]
```diff
@@ -242,6 +249,126 @@ namespace crypto {
     return sc_isnonzero(&c) == 0;
   }
 
+  void crypto_ops::generate_tx_proof(const hash &prefix_hash, const public_key &R, const public_key &A, const public_key &D, const secret_key &r, signature &sig) {
+    boost::lock_guard<boost::mutex> lock(random_lock);
```
Comment (stoffu) [inline src/crypto/crypto.cpp]: It's there just because I wrote this function by first copying code from `generate_signature()`. I see 3 other similar uses of `boost::lock_guard` in crypto.cpp. Shall I change those as well in this PR, or make a separate PR, or leave them?

File: src/simplewallet/simplewallet.cpp (hunk 6 above)
Comment (stoffu) [inline src/simplewallet/simplewallet.cpp]: Again, this exists because I copied the code from `check_tx_key()`, and I see 5 other occurrences of it in simplewallet.cpp. First, how should it be changed? Should it print an error message by `fail_msg_writer()` and return true, or should it throw like `throw std::runtime_error("simple_wallet null wallet");` as done in `get_daemon_blockchain_height()`? Second, should I also change other parts in simplewallet.cpp doing the same?

File: src/simplewallet/simplewallet.cpp (hunk 7 above)
Comment (stoffu) [inline src/simplewallet/simplewallet.cpp]: This was also copied from `check_tx_key()`. Where exactly should it be moved to? I have only a vague understanding that `LOCK_IDLE_SCOPE();` is used when the program accesses some data stored in the wallet from various threads concurrently.

Actually this function (and also `check_tx_key()`) should be completely independent of the wallet data (i.e. they could be implemented in the daemon), so this `LOCK_IDLE_SCOPE();` as well as the above `assert(m_wallet);` could be dropped altogether. Am I correct?

File: src/simplewallet/simplewallet.cpp (hunk 4 above)
Comment (stoffu) [inline src/simplewallet/simplewallet.cpp]: Yes. I did this precisely because `null_skey` isn't defined in cryptonote_basic.h. The definition would have been like:

```
const static crypto::public_key null_pkey = AUTO_VAL_INIT(null_pkey);    // this exists currently
const static crypto::secret_key null_skey = AUTO_VAL_INIT(null_skey);
```

Should I add this line to cryptonote_basic.h?
Or I could use `rct::I` instead.

File: src/simplewallet/simplewallet.cpp (hunk 1 above)
Comment (stoffu) [inline src/simplewallet/simplewallet.cpp]: ok

File: src/simplewallet/simplewallet.cpp (hunk 2 above)
Comment (stoffu) [inline src/simplewallet/simplewallet.cpp]: ok

File: src/simplewallet/simplewallet.cpp (hunk 3 above)
Comment (stoffu) [inline src/simplewallet/simplewallet.cpp]: ok

File: src/simplewallet/simplewallet.cpp (hunk 5 above)
Comment (stoffu) [inline src/simplewallet/simplewallet.cpp]: ok

File: src/crypto/crypto.cpp (hunk 8 above)
Comment (moneromooo-monero) [inline src/crypto/crypto.cpp]: If you want to change those, either a separate PR, or the same PR in a different commit.

File: src/simplewallet/simplewallet.cpp (hunk 6 above)
Comment (moneromooo-monero) [inline src/simplewallet/simplewallet.cpp]: I think all those functions return errors, and catch any exception that wallet2 might throw. So do the same (assuming I remember right, I did not check right now). If you want to change the other parts, another commit seems fine.

File: src/simplewallet/simplewallet.cpp (hunk 7 above)
Comment (moneromooo-monero) [inline src/simplewallet/simplewallet.cpp]: LOCK_IDLE_SCOPE ensures wallet2 is not currently syncing the blockchain, and locks it from doing so until the end of the scope. Basically, it seems pointless to wait/lock if you're going to do through an early error path anyway. As I said, optional minor things.

File: src/simplewallet/simplewallet.cpp (hunk 4 above)
Comment (moneromooo-monero) [inline src/simplewallet/simplewallet.cpp]: Hmm. I suppose not having null_skey can prevent dumb mistakes. So it's fine as is, if you're certain the auto init sets it to zero always.

File: src/crypto/crypto.cpp (hunk 8 above)
Comment (stoffu) [inline src/crypto/crypto.cpp]: ok

File: src/simplewallet/simplewallet.cpp (hunk 4 above)
Comment (stoffu) [inline src/simplewallet/simplewallet.cpp]: OK, I'd just use `rct::I`.

File: src/simplewallet/simplewallet.cpp (hunk 6 above)
Comment (stoffu) [inline src/simplewallet/simplewallet.cpp]: OK, I'll do another commit.

File: src/simplewallet/simplewallet.cpp (hunk 7 above)
Comment (stoffu) [inline src/simplewallet/simplewallet.cpp]: OK, so I don't see a need to use `LOCK_IDLE_SCOPE` here and omitted it. Also in a separate commit I removed it from `check_tx_key()` for the same reason.

**Comment by stoffu** [comment]
Thank you @moneromooo-monero for reviewing! I hope my revision is satisfactory. Also final checking from @luigi1111 is greatly appreciated.

**Comment by moneromooo-monero** [comment]
Please squash commits which only fix something done in a prior commit in the same PR.

File: src/crypto/crypto.cpp [hunk 9]
```diff
@@ -221,7 +220,10 @@ namespace crypto {
 #endif
     buf.h = prefix_hash;
     buf.key = pub;
-    random_scalar(k);
+    {
+      boost::lock_guard<boost::mutex> lock(random_lock);
+      random_scalar(k);
+    }
```
Comment (moneromooo-monero) [inline src/crypto/crypto.cpp]: I'd move the lock to a new random_scalar function, and rename the current random_scalar to random_scalar_not_thread_safe, since it is not, and that was a source of bugs earlier.

Comment (stoffu) [inline src/crypto/crypto.cpp]: ok

**Comment by stoffu** [comment]
> commits which only fix something done in a prior commit in the same PR

Sorry but I couldn't understand your intention. Which commits should I squash? I think they're unrelated to each other.

**Comment by moneromooo-monero** [comment]
Ah, those are moving things in other places, not in your new code, I missed that. All good then, thanks.
